Allowing OSPRay to export modern CMake targets during installation (and build)

wilsonCernWq opened the pull request:
Hi All, 

#### brief
 
I just made some changes in OSPRay's cmake system so that we can export ospray's targets during building and installation, however, one can still use the osprayConfig.cmake as before since nothing was deleted.

#### what has been changed

Basically I add export targets for all the libraries created by the `ospray_create_library` macro under the namespace `ospray::`. Therefore a 3rd party application can now easily link to an ospray target using one line `target_link_libraries(<target> ... ospray::ospray ospray::ospray_common ...)`, without the need of worrying about, for example, setting up include targets. Application can now also link against module libraries if it is desired. I also made some external libraries `PRIVATE` such as `glfw` since we don't want to export them through ospray.

#### note
I have ran some test for those changes, but certainly I don't think I have tested all the cases. Let me know if there are things I should fix :)

Best

Qi

carsonbrownlee commented:
Thank you Qi!  This was an interesting read on the topic for me: https://pabloariasal.github.io/2018/02/19/its-time-to-do-cmake-right/ .    If I follow this correclty, you have created an export target for every generated library, but we dont' want users using our internal libraries. For example you mention the glfw dependancy, the app libraries are not meant to be used externally period. @jeffamstutz thoughts?

wilsonCernWq commented:
Hi Carson

Yes, the good things about target is, a target can carry lots of informations (e.g. include directories, compile definitions and even linkages to other targets) that we have to manually setup before. 

>  If I follow this correclty, you have created an export target for every generated library

Yes, from the perspective of CMake, if a target is built internally in the project, the target should be exported. In fact one exported target implies there is one library file associated to it.

> But we dont' want users using our internal libraries.

It may cause problems if such an internal library is linked by any targets we want to export. For example, if we only allow users to use `ospray::ospray`, all the internal dependencies (those dependencies cannot be found outside ospray) for `ospray::ospray` should be exported. Which is why it is okay to ser `glfw` target `PRIVATE` since applications can supply glfw themselves if they are building glfw applications. However we can certainly put those libraries under a different namespace, say `_ospray_::` and "hide" them from users ... Users won't use them if they don't manually include them ... But they need to be exported because other targets depends on it, and those important targets need to know where to find those "hidden" libraries on disk.

However it is fine if these internal libraries are not linked by exported targets such as `ospray::ospray` or `ospray::ospray_common`, or if those libraries are linked by executables ... There are several ways to exclude them from the export list:

* we can certainly call cmake functions `add_library` and `target_link_directories` directly and don't call make them installed/exported.

* we can also add parameters for `ospray_create_libraries` and don't add them in the export list `OSPRAY_TARGET_LIST`

Qi

carsonbrownlee commented:
> For example, if we only allow users to use ospray::ospray, all the internal dependencies (those dependencies cannot be found outside ospray) for ospray::ospray should be exported

Yes, but my concern is that you have made a list of targets which include libraries built for our apps only.  naming them to another namespace such as osprayinternal:: may work, but it sounds better to me if they aren't exported targets at all.

wilsonCernWq commented:
> Yes, but my concern is that you have made a list of targets which include libraries built for our apps only. naming them to another namespace such as osprayinternal:: may work, but it sounds better to me if they aren't exported targets at all.

for those targets, we need a new parameter in our cmake macro to distinguish them. I can add that functionality, but I am not entirely sure which libraries are needed and which are not.

wilsonCernWq commented:
Hi Carson

I just added the support for explicitly turn-off export for target. By adding `EXCLUDE_FROM_EXPORT` parameter to `OSPRAY_CREATE_LIBRARY` will automatically exclude targets from exporting. I also added this flag for all the targets under the `app` folder

Qi

Twinklebear commented:
Cool, thanks Qi! Modern CMake exported targets for OSPRay is something that I've been wanting as well. I'll have to test this a bit, but for now I have one smaller comment. I think it would be better to have libraries opt-in to be exported instead of opt-out, as you have now. This would make it easier for writing modules (which shouldn't be exported as targets to link against), since the end-user wouldn't need to change their code. Then in the rarer case we do want a module to be exported (e.g., maybe the ISPC module) we can add a `EXPORT_LIBRARY` in the create library call.

wilsonCernWq commented:
> I think it would be better to have libraries opt-in to be exported instead of opt-out, as you have now.

I have changed that ... so now you just need to add `EXPORT` somewhere in the `CREATE_OSPRAY_LIBRARY` call to export a target.

jeffamstutz commented:
Hi Qi,

Thanks for taking the time to work this out! We are very focused on SIGGRAPH right now, but after that I'd be happy to take this and fully "modernize" OSPRay's CMake....it's been a long time coming! It helps to know that an external collaborator would benefit from modernizing.

jeffamstutz commented:
Also FYI, I'll probably merge this into a different branch I can work on to verify our CI environment still works.

wilsonCernWq commented:
Certainly Jeff! Close this PR whenever you think it is okay :D

wilsonCernWq commented on components/ospcommon/cmake/macros.cmake:
```diff
@@ -206,6 +223,24 @@ MACRO(OSPRAY_INSTALL_LIBRARY name component)
+  IF ("${export}" STREQUAL "TRUE")
+    # Export targets to install directory
+    # Note (Qi): this creates one file per target, thus we have to include all of them
+    #   when compiling ospray related applications
```
I know we can append target to one single export file if we export to the build tree. But it seems we can not do that if we want to export to the install directory. Is there a reason for that? 

Actually the reason I created multiple export files for each target was because ospray will call install multiple times for each target, I didn't want to change it too dramatically ... But yeah, one single export file will be neat if possible ...

wilsonCernWq commented on cmake/ospray_cmake_config/osprayConfig.cmake.in:
```diff
@@ -243,3 +243,16 @@ set(OSPRAY_USE_FILE ${OSPRAY_CMAKE_ROOT}/osprayUse.cmake)
+    # include all the target files
+    include("${OSPRAY_CMAKE_ROOT}/ospray-${f}-targets.cmake")
+    # Note (Qi): this is actually not necessary, but this prevents compilation errors when someone
+    #   used 'include_directories' by accidents.
```
Actually I was wrong in the note. The only reason I added OSPRAY_INCLUDE_DIRS explicitly was because in ospray's implementation, `include_directories` is widely used instead of `target_include_directories` ... I wasn't sure if ospray wants such a big change ... @jeffamstutz

wilsonCernWq commented on apps/CMakeLists.txt:
```diff
@@ -31,9 +31,11 @@ OPTION(OSPRAY_APPS_UTILITIES "Build utility apps" ON)
-  ADD_SUBDIRECTORY(common/ospapp)
-  ADD_SUBDIRECTORY(common/sg)
+  # Note (Qi): here we have to build tfn_lib, sg and ospapp IN ORDER
+  #   because otherwise we will have a wrong target export order ...
```
Yes, currently I don't know how to do one export for all ospray targets ...

mathstuf replied:
Instead of `EXPORT ospray-${name}`, just do `EXPORT ospray`.

wilsonCernWq commented on components/ospcommon/cmake/macros.cmake:
```diff
@@ -180,23 +180,40 @@ ENDMACRO()
-    LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
+MACRO(OSPRAY_INSTALL_LIBRARY name component export)
+
+  IF ("${export}" STREQUAL "TRUE")
```
I don't know why but `if(export)` doesn't work here ... `export` is a macro argument whose value is written by this [line](https://github.com/ospray/ospray/pull/263/files#diff-803f304f6b6cd47bf07abc0d07850bb9R307).

mathstuf replied:
Oh, yeah, that makes sense. Macros are kind of dumb, but doing `macro` → `function` is a separate PR.

wilsonCernWq commented on components/ospcommon/cmake/macros.cmake:
```diff
@@ -180,23 +180,40 @@ ENDMACRO()
       COMPONENT devel
-  )
+      )
+  ELSE()
```
That looks cool, thank you :D

wilsonCernWq commented:
Hi All, @jeffamstutz @carsonbrownlee @Twinklebear 

I added some updates after following Ben's helpful advices @mathstuf 
Comparing to the previous update, I added the following thing:

* I added install rules for `ospray` and `ospray_common`'s include path as you can see [here](https://github.com/ospray/ospray/pull/263/files#diff-7fa5433b42db3a16a9b8c0358b996874R481) and [here](https://github.com/ospray/ospray/pull/263/files#diff-a0780d53a93f877ea5dde0b5d13e3f7bR95). I set only the install rule because I don't want to break ospray's `include_directories` machanism that has been already widely used. Also I didn't add this rules to all targets ... well certainly we should, ... but I think if a user is using, say, `ospray_module_xxx` without linking against to `ospray`, it is a problem already ... 

* I also removed `snappy` from the public linkage. I think a 3rd party app should supply `snappy` itself if needed ... see [code1](https://github.com/ospray/ospray/pull/263/files#diff-d2d1ab8f14ee1dea69702f8e6a6538baR74) and [code2](https://github.com/ospray/ospray/pull/263/files#diff-fb09a48f58658a585dbddf5afcef13f7R36)

* and I used a global property to hold the export list instead of a cache ...

jeffamstutz commented:
UPDATE: I'm almost finished with a complete refactor of OSPRay's CMake, where I think I have everything listed in this PR accounted for. I'll close this when it is merged and pushed to ```devel```.

mathstuf commented:
Cool. I'll test it on Linux with VTK once it's merged.

jeffamstutz commented:
Currently this is implemented in both ```devel``` and ```release-1.8.x``` and will be a part of the upcoming ```v1.8.0``` release.
